testing/testing_dexterity.rst

```diff
@@ -0,0 +1,188 @@
+
+
+Excercise
++++++++++
+
+Try to change the permissions for our content-type and set that only Manager can add it.
+Then fix the test for contributor role and create a new test for the Manager.
+
+..  admonition:: Solution
+    :class: toggle
+
+    In ``rolemap.xml``:
+
+    .. code-block:: xml
+
+        <permission name="plonetraining.testing: Add TestType" acquire="True">
+            <role name="Manager"/>
+        </permission>
+
+    In test case file:
+
+    .. code-block:: python
+
+        from AccessControl.unauthorized import Unauthorized
+        ...
+
+        def test_ct_test_type_adding(self):
+            setRoles(self.portal, TEST_USER_ID, ['Manager'])
+            ...
+
+        def test_ct_test_type_adding_contributor(self):
+            setRoles(self.portal, TEST_USER_ID, ['Contributor'])
+            self.assertRaises(
+                Unauthorized,
+                api.content.create,
+                container=self.portal,
+                type='TestType',
+                id='test_type',
+            )
+
+
+These are ``integration`` tests because we are not testing the browser integration.
+
+We could try to create a ``functional`` test to test how our content-type creation works on browser.
+
+Let's create a new test class in the same file like this:
+
+.. code-block:: python
+
+    from plone.testing.z2 import Browser
+    from from plonetraining.testing.testing import PLONETRAINING_TESTING_FUNCTIONAL_TESTING
+    ...
+
+    class TestTypeFunctionalTest(unittest.TestCase):
+
+        layer = PLONETRAINING_TESTING_FUNCTIONAL_TESTING
+
+        def setUp(self):
+            app = self.layer['app']
+            self.portal = self.layer['portal']
+            self.request = self.layer['request']
+            self.portal_url = self.portal.absolute_url()
+
+            # Set up browser
+            self.browser = Browser(app)
+            self.browser.handleErrors = False
+            self.browser.addHeader(
+                'Authorization',
+                'Basic %s:%s' % (SITE_OWNER_NAME, SITE_OWNER_PASSWORD,)
+            )
+
+        def test_add_test_type(self):
+            self.browser.open(self.portal_url + '/++add++TestType')
+            self.browser.getControl(
+                name="form.widgets.IBasic.title"
+            ).value = "Example content"
+            self.browser.getControl("Save").click()
+
+            self.assertEqual(
+                "Example content", self.portal['example-content'].title
+            )
+
+        def test_view_test_type(self):
+            setRoles(self.portal, TEST_USER_ID, ['Manager'])
+            self.portal.invokeFactory(
+                "TestType",
+                id="example-content",
+                title="Example content",
+                description="This is a description",
+            )
+
+            import transaction
+
+            transaction.commit()
```

Comment:
There should be a note about transactions and commits, why those are needed here and when its a bad idea to use them in tests.

Reply:
i should have write some infos about it now. Try to see if it's cleaner now